### DISCUSSION TITLE
Generic Error | The scanner was already started. Call stop() before calling start() again

### DISCUSSION THREAD

**Discussion by rodnnnney** [juliansteenbakker/mobile_scanner#1152] [Q&A]
The scanner works and initializes when the user navigates to page. However, when the user switches to another page and back the page where I have my scanner.

 It throws this error: 
`Generic Error | The scanner was already started call stop() before calling start() again. `

To try to clear this error I hot restarted but the error just changed to 
`Generic Error | Called start() while already started()!`
 Can the package not handle this and just ignore the start call if it's already started instead of throwing an error?

Here is the initialization portion:
  
```

late MobileScannerController controller;
StreamSubscription<Object?>? _subscription;
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
    _initializeController();
  }

  Future<void> _initializeController() async {
    controller = MobileScannerController(
      detectionSpeed: DetectionSpeed.noDuplicates,
      facing: CameraFacing.back,
      torchEnabled: false,
      formats: [BarcodeFormat.qrCode],
    );
    await _startScanner();
  }

```

Here is the part where I thought it handles how the scanner should respond if there was a page change via didChangeAppLifecycleState

```

Future<void> _startScanner() async {
    if (!_isScannerRunning) {
      try {
        await controller.start();
        _isScannerRunning = true;
        _subscription = controller.barcodes.listen((barcode) {
          print('Barcode found: ${barcode.barcodes.first}');
        });
      } catch (e) {
        print('Error starting scanner: $e');
      }
    }
  }

  Future<void> _stopScanner() async {
    if (_isScannerRunning) {
      await _subscription?.cancel();
      _subscription = null;
      controller.stop();
      _isScannerRunning = false;
    }
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (!controller.value.isInitialized) {
      return;
    }
    switch (state) {
      case AppLifecycleState.resumed:
        _startScanner();
        break;
      case AppLifecycleState.inactive:
        _stopScanner();
        break;
      case AppLifecycleState.paused:
        _stopScanner();
        break;
      case AppLifecycleState.detached:
        _stopScanner();
        break;
      default:
        _stopScanner();
        break;
    }
  }

@override
  Future<void> dispose() async {
    WidgetsBinding.instance.removeObserver(this);
    await _stopScanner();
    await controller.dispose();
    pageController.dispose();
    super.dispose();
  }

```

Here is the actual controller page

```
@override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: PageView(
        controller: pageController,
        onPageChanged: (index) async {
          await _stopScanner();
          if (index == 0) {
            await Future.delayed(const Duration(milliseconds: 500));
            if (mounted) {
              await _startScanner();
            }
          }
        },
        children: [
          _BarcodeScannerPage(
            controller: controller,
          ),
        ],
      ),
    );
  }
```

Does anyone know why this is not handling the issue as I thought it would be? 

I thought it could be related to 
```
class _BarcodeScannerPageViewState extends State<BarcodeScannerPageView>
    with RouteAware
```
Since we are popping pages, I thought it could be the fact that WidgetsBindingObserver didn't popping and navigating new pages onto the context, but it still called start at the init of the page.

However that implementation didn't work either. Has anyone else ran into similar errors and found a solution?

**Comment by dafayape** [comment]
The root cause is that `_isScannerRunning` is a manually-tracked flag that can get out of sync with the controller's actual state across `await` boundaries. Between `await controller.start()` starting and `_isScannerRunning = true` executing, another call path (your `onPageChanged` handler firing alongside `didChangeAppLifecycleState`, which can both trigger in quick succession during a page transition) can slip through the `if (!_isScannerRunning)` guard before the flag updates — so you get two concurrent `start()` calls.

Two fixes that solve this reliably:

1. **Use the controller's own state instead of a manual bool.** Check `controller.value.isRunning` instead of your own `_isScannerRunning` — it reflects the plugin's actual internal state, not a copy that can drift.
2. **Wrap start()/stop() in try-catch and treat "already started/stopped" as a no-op**, since the plugin throws rather than silently ignoring redundant calls:
   ```dart
   Future<void> _startScanner() async {
     try {
       await controller.start();
     } on MobileScannerException catch (e) {
       if (e.errorCode != MobileScannerErrorCode.controllerAlreadyStarted) rethrow;
     }
   }
   ```

Also worth removing the redundant `onPageChanged` start/stop calls if `didChangeAppLifecycleState` already covers it — having two independent triggers calling the same start/stop logic is exactly what's racing here.
